File: Array/75-sortColors/go/75.go
package main

func sortColors(nums []int)  {
    p0, p1 := 0, 0
	for i := range(nums) {
		if nums[i] == 1 {
			nums[p1], nums[i] = nums[i], nums[p1]
			p1++
		} else if nums[i] == 0 {
			nums[p0], nums[i] = nums[i], nums[p0]
			if p0 < p1 {
				nums[p1], nums[i] = nums[i], nums[p1]		
			}
			p0++
			p1++
		}
	}
}

File: Array/75-sortColors/75.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        // 双指针
        int p0 = 0;  // 负责交换 0
        int p1 = 0;  // 负责交换 1
        for (int i = 0; i < nums.size(); ++i) {
            if (nums[i] == 1) {
                swap(nums[p1], nums[i]);
                ++p1;
            } else if (nums[i] == 0) {
                swap(nums[p0], nums[i]);
                if (p0 < p1) {  // 则刚才交换了一个 1 到 nums[i]
                    swap(nums[p1], nums[i]);  // 把 1 换回去
                }
                ++p0, ++p1;
            }
        }
    }
};

int main() {
    vector<int> nums = {0, 1};
    Solution obj;
    obj.sortColors(nums);
    return 0;
}